Test recorder (BETA) draft (#115)
mulesoft/docs-munit: modules/ROOT/pages/test-recorder.adoc

> = Test Recorder in Studio (Beta)

Author: @kip-mulesoft This is an idea of the landing page for the Test recorder.
It might be a good fit for the "Learn More" link in the wizard. 

I'll add more information about the test recorder itself. Do you think it would be good to list the Limitations of the tool being beta?

Reviewer: I think this is the best place to put limitations, especially if this is where we point our learn more link.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

The test recorder enables you to record a processing flow and then configure a unit test based on the captured event.

> The test recorder is an open beta feature and is compatible with Anypoint Studio 7.x.x and later, and MUnit 2.2.1 and later.

Reviewer: This is not correct. At the very least, we need MUnit 2.2.2 and Mule 4.2.2, and I haven't checked the recorder with any Studio version older than 7.4.0. Later I will do some testing and send you which versions should be placed here.

Reviewer: Also, you need a patched version of 4.2.2, so you will need to pull the patch from the updatesite that has the required dependencies on it. The users should be prompted as soon as the new version becomes available and they might not even realise that they have already fetched the recorder dependencies, but we should still notify them just in case there is any issues.

Reviewer: Now that we are nearing release, supported versions have already been defined. It will be needing at least MUnit 2.2.4 (which comes bundled with MUnit's Studio plugin, so this should be transparent to the user), the patched Mule 4.2.2, and Studio 7.4.x.

Reviewer: Also, it might be good to note that these versions will definitely change from BETA to GA.

> The test recorder is an open beta feature and is compatible with Anypoint Studio 7.x.x and later, and MUnit 2.2.1 and later.

Reviewer 2: ```suggestion
The test recorder is an open beta feature that is compatible with Anypoint Studio 7.x.x and later and MUnit 2.2.1 and later.
```
beta, Beta, or BETA?

= Test Recorder in Studio (Beta)
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

The test recorder enables you to record a processing flow and then configure a unit test based on the captured event.

> The test recorder captures real data as it goes through your application in Studio. By capturing the flow execution, MUnit can automatically generate an MUnit test so that you can configure the necessary mocks and assertions for your test. See xref:test-recorder-examples.adoc[Recording a Test from an Exchange Example] for a fully working example.

Reviewer 2: ```suggestion
The test recorder captures real data as it goes through your application in Anypoint Studio. By capturing the flow execution, MUnit can automatically generate an MUnit test so that you can configure the necessary mocks and assertions for your test. See xref:test-recorder-examples.adoc[Recording a Test from an Exchange Example] for a fully working example.
```

[NOTE]

> The test recorder is an open beta feature.

Reviewer 2: This is in the topic title; do we need to repeat it here?

== Compatibility

> The following table describes the required product versions to use the test recorder. These versions are for the beta feature and are likely to change for the GA (general availability) release.

Reviewer 2: ```suggestion
The following table describes the product versions required to use the test recorder. These versions are for the beta feature and are likely to change for the GA (general availability) release.
```

[%header%autowidth.spread,cols="a,a"]
|===
| Product | Version
| Anypoint Studio

> | 7.4.0 and later.

Reviewer 2: ```suggestion
| 7.4.0 and later
```

| MUnit
| 2.2.4

| MUnit Anypoint Studio Plugin
| 2.4.2

| Mule runtime engine
| 4.2.2

|===

== Limitations

> The test recorder has the following limitations for creating MUnit tests:

Reviewer 2: ```suggestion
The test recorder is limited in the following ways when you create MUnit tests:
```

> * You cannot create tests for flows that have mule errors either raised inside the flow or already inside the incoming event. Not even if they are handled by an *on error continue* error handler.

Reviewer 2: ```suggestion
* You cannot create tests for flows with Mule errors raised inside the flow or already existing in the incoming event, not even if they are handled by an *on error continue* error handler.
```

> * When the recorded flow execution ends, the application is killed. This causes the result of the execution to never reach its destination. For example, an HTTP listener will not respond to the incoming request. This does not mean that the recording failed.

Reviewer 2: Limit list items and steps to one sentence maximum, to enhance scannability (which is the purpose of a list).
Also, remember that the items in this list are supposed to be limitations, so begin each with what the real limitation is: in this case, I think the limitation is that the result of the flow does not reach its destination?
This is just my guess; please reword for accuracy, as appropriate:
```suggestion
* A recorded flow execution ends successfully, but the result does not reach its destination because the application is killed. 
```

> * Don't configure Spy or Assert processors to assert values for random data, time-dependent information (such as timestamps), or values resulting from parallel processes. Because those values change in every execution, your validations will fail every time your test runs.

Reviewer 2: Again, list the limitation:
```suggestion
* Your validations fail every time your test runs if you configure Spy or Assert processors to assert values for random data, time-dependent information (such as timestamps), or values resulting from parallel processes, because those values change in every execution.
```

> * Don't mock values resulting from parallel processes because the mixture of real and mocked data will compromise the execution of the processors that follow in your test.

Reviewer 2: ditto:
```suggestion
* Mocking values resulting from parallel processes causes a mixture of real and mocked data that compromises the execution of the processors that follow in your test.
```

> * Although the recorder supports data iteration in the flow, such as recursivity or loops, it does not support cases where the structure of the data being tested changes inside said iteration.

Reviewer 2: ```suggestion
* Although the recorder supports data iteration in the flow, such as recursivity or loops, it does not support cases in which the structure of the data being tested changes inside the iteration.
```

> * The recorder does not support mocking a message before or inside a Foreach processor.

Reviewer 2: ```suggestion
* The recorder does not support mocking a message before or inside a foreach processor.
```

The test recorder enables you to record a processing flow and then configure a unit test based on the captured event.

The test recorder captures real data as it goes through your application in Studio. By capturing the flow execution, MUnit can automatically generate an MUnit test so that you can configure the necessary mocks and assertions for your test. See xref:test-recorder-examples.adoc[Recording a Test from an Exchange Example] for a full working example.

[NOTE]
The test recorder is an open beta feature.

== Compatibility

The following table describes the required product versions to use the test recorder. These versions are for the beta feature and are likely to change for the GA (general availability) release.

[%header%autowidth.spread,cols="a,a"]
|===
| Product | Version
| Anypoint Studio
| 7.4.0 and later.

| MUnit

> | 2.2.4

Reviewer: This is the runtime version of MUnit, but what you need to get is the `MUnit Studio Plugin` version 2.4.2, which will come bundled with the `MUnit Runtime` 2.2.4 and the `MUnit Test Recorder` 1.0.0.

Author: Then I'd suggest adding MUnit Anypoint Studio Plugin 2.4.2 to this list. (this is how it shows in the update site)

| Mule runtime engine
| Patched version of 4.2.2. +

> See xref:runtime-patching.adoc[Patching the Embedded Mule Runtime Engine Version] to learn how to apply this patched version to your project.

Reviewer: This is not how you apply this patched version. That defines the version used while running the MUnit tests, but the recorder happens during the app execution, not the test execution. What you need to do is to go to `Install new software`, in the `Work with` drop down menu select `Anypoint Studio Enterprise Runtimes`, and then select `Mule Server Runtime 4.2.2 EE`. This should install the latest 4.2.2 runtime available.

Author: Got it. Wrong link then.

|===

== Limitations

The test recorder has the following limitations for creating MUnit tests:

* You cannot create tests for flows that have mule errors either raised inside the flow or already inside the incoming event. Not even if they are handled by an *on error continue* error handler.
* When the recorded flow execution ends, the application is killed. This causes the result of the execution to never reach its destination. For example, an HTTP listener will not respond to the incoming request. This does not mean that the recording failed.

> * The test recorded does not support creating tests for anything that uses parallelism, random data, or time dependent information (for example, timestamps).

Reviewer: The test recorder, not the test recorded. And it is not that it is not supported, but rather that if configured incorrectly (e.g. try to assert a timestamp), then the test will always fail (e.g. the timestamp will always change). This is not an issue of the recorder per se, but rather illogic testing, regardless of the tool.

Author: If it's not a limitation of the feature, I wouldn't suggest adding it here. Although I see why it would be useful information for new users. 
I'd say:

> You should not write assertions  for anything that uses parallelism, random data, or time-dependent information (e.g. timestamps). Because those values will change in every execution, your assertions will fail every time your test runs.

Or something like that 🤔

Reviewer: Exactly my point. The idea of the recorder is to bring in new users (as well as helping existing ones), and the tool will not stop you from doing these things, so putting it here helps fill that gap. What I would add to your suggestion is that both spies and assertions will fail in all these cases (spies are just assertions in the middle of the execution), but mocks will only fail with parallelism, and actually mocking random data or time-dependent information is actually something that would help make the execution more consistent and therefore easier to test. Not sure how to put all that mess into a coherent sentence, though. Maybe something like this?

> You should not configure spies or assertions for anything that uses parallelism, random data, or time-dependent information (for example, timestamps), because those values will change in every execution, and those actions will fail every time your test runs. For similar reasons, you should not mock anything that uses parallelism, because there will be a mix of real and mocked data that will break anything that follows.

Author: I'd go with an imperative voice for this sentence:

> * Don't configure Spy or Assert processors to assert values for random data, time-dependent information (such as timestamps), or values resulting from parallel processes. Because those values change in every execution, your validations will fail every time your test runs. 
>
> * Don't mock values resulting from parallel processes because the mixture of real and mocked data will compromise the execution of the processors that follow in your test.

Reviewer: Looks great!

* Although the recorder supports data iteration in the flow, such as recursivity or loops, it does not support cases where the structure of the data being tested changes inside said iteration.

> * The recorder does not support mocking a message before or inside a Foreach processor. +

Reviewer: Shouldn't we link the documentation of said test case (https://docs.mulesoft.com/munit/latest/mock-before-after-foreach-cookbook)? Otherwise users unfamiliar with the tool might be more confused after reading this because they do not know if this affects them or not.

Author: I think it would be better to add that link to the See Also section right below this sentence.
